tools/generators/target_build_settings/src/Generator/ParseTransitiveSwiftDebugSettings.swift

```diff
@@ -0,0 +1,145 @@
+import Foundation
+import OrderedCollections
+import ToolCommon
+
+extension Generator {
+    struct ParseTransitiveSwiftDebugSettings {
+        private let callable: Callable
+
+        /// - Parameters:
+        ///   - callable: The function that will be called in
+        ///     `callAsFunction()`.
+        init(callable: @escaping Callable = Self.defaultCallable) {
+            self.callable = callable
+        }
+
+        /// Reads transitive Swift debug settings to set Swift debug settings
+        /// for the current target.
+        func callAsFunction(
+            _ transitiveSwiftDebugSettingPaths: [URL],
+            clangArgs: inout [String],
+            frameworkIncludes: inout OrderedSet<String>,
+            onceClangArgs: inout Set<String>,
+            swiftIncludes: inout OrderedSet<String>
+        ) async throws {
+            try await callable(
+                /*transitiveSwiftDebugSettingPaths:*/
+                    transitiveSwiftDebugSettingPaths,
+                /*clangArgs:*/ &clangArgs,
+                /*frameworkIncludes:*/ &frameworkIncludes,
+                /*onceClangArgs:*/ &onceClangArgs,
+                /*swiftIncludes:*/ &swiftIncludes
+            )
+        }
+    }
+}
+
+// MARK: - ParseTransitiveSwiftDebugSettings.Callable
+
+extension Generator.ParseTransitiveSwiftDebugSettings {
+    typealias Callable = (
+        _ transitiveSwiftDebugSettingPaths: [URL],
+        _ clangArgs: inout [String],
+        _ frameworkIncludes: inout OrderedSet<String>,
+        _ onceClangArgs: inout Set<String>,
+        _ swiftIncludes: inout OrderedSet<String>
+    ) async throws -> Void
+
+    static func defaultCallable(
+        _ transitiveSwiftDebugSettingPaths: [URL],
+        clangArgs: inout [String],
+        frameworkIncludes: inout OrderedSet<String>,
+        onceClangArgs: inout Set<String>,
+        swiftIncludes: inout OrderedSet<String>
+    ) async throws {
+        for url in transitiveSwiftDebugSettingPaths {
+            var iterator = url.lines.makeAsyncIterator()
+
+            guard let rawClangArgsCount = try await iterator.next() else {
+                throw PreconditionError(
+                    message: url.prefixMessage("Missing clang args count")
+                )
+            }
+            guard let clangArgsCount = Int(rawClangArgsCount) else {
+                throw PreconditionError(message: url.prefixMessage("""
+Clang args count "\(rawClangArgsCount)" was not an integer
+"""))
+            }
+            argsLoop: for index in (0..<clangArgsCount) {
+                guard let arg =
+                    try await iterator.next()?.nullsToNewlines
+                else {
+                    throw PreconditionError(message: url.prefixMessage("""
+Too clang args. Found \(index), expected \(clangArgsCount)
```

Comment:
"Too few"? IIUC this handles the iterator finishing earlier than expected correct?

Reply:
Fixed in #2708.